src/groups/group.jl

```diff
@@ -1121,6 +1200,35 @@ end
 direction_and_side(::GroupExponentialRetraction{D}) where {D} = D()
 direction_and_side(::GroupLogarithmicInverseRetraction{D}) where {D} = D()
 
+function log(::TraitList{<:IsGroupManifold}, G::AbstractDecoratorManifold, p, q)
```

Comment:
In my PR I've restricted these `log` and `exp` methods to semidirect product groups with with left-invariant tangent vector representation. On many groups we have optimized implementation on the manifold so I'd prefer to have that to be picked up first before this relatively generic fallback.

Reply:
The solution here is to call the exponential/logarithms defined in `groups/connections.jl`. It's the only acceptable `exp` and `log` on a group manifold. This way, it will work for *all* groups, regardless of the tangent vector representation. Should I fix this in this PR?

Reply:
I know your solution is correct, this is mostly about performance (and a bit about consistency, see the next paragraph). Substantial amount of work has been put to make practically relevant calls fast but unfortunately it's not gathered in an easy to run benchmark suite, so verifying what could have been slowed down and fixing issues would take me hours. So, unless you know of a case where my solution gives a wrong result for an existing group, I'd prefer to keep it like that for now.

Additionally, considering your implementation here makes me rethink our entire group -> manifold pass-through and it's not really something I'd like to do in the near future. If we don't pass exp and log to the manifold, why do we pass vector transports, inner product, retractions, inverse retractions, Riemann tensor, affine connection and so on? Maybe the current solution is not ideal but reworking it would take a very long time and I just don't see enough benefits.

> Should I fix this in this PR?

Since I've already merged your branch, it would be easier for me if you made any further fixes on top of the branch from #732 .

Reply:
vector transport and retractions (and their inverses) are still valid, at least after turning back to tangent vectors in tangent spaces (not 100% sure about Lie algebra things); but it might be beneficial to maybe not pass through exp/log. But I agree that that is something that has to be considered carefully and in the near future I do not have the capacity for that.

Reply:
Thank you for your feedback and explanations! I hope I did not completely misunderstand.

1. Before my PR, the `exp` that was used was not necessary the group exponential (the one in `groups/connections.jl`). For instance, for semidirect products, the exponential of the underlying *product* manifold was used, which was wrong, because it was not invariant. My implementation fixed that, but was also flawed since it assumed left invariant representation. This is easily fixed as I mentioned above. I'd be happy to do that.

2. I'm not sure I understand the implications of what you call the "pass through" here... I'd be happy to be enlightened.
As far as speed is concerned, `exp` is always an expensive operation, and I doubt that the overhead of function calls makes any difference. Am I wrong to assume that?

3. The only optimized `exp` that I can find in the code is that of `GeneralLinear`. It comes from a choice of left invariant metric, and I'm pretty sure it is not the same as the invariant exponential in `groups/connections.jl`. (So that optimized exponential is *not* invariant, unless I am mistaken) Have I missed any other optimized `exp` implementation?

Reply:
> 1. Before my PR, the `exp` that was used was not necessary the group exponential (the one in `groups/connections.jl`). For instance, for semidirect products, the exponential of the underlying _product_ manifold was used, which was wrong, because it was not invariant. My implementation fixed that, but was also flawed since it assumed left invariant representation. This is easily fixed as I mentioned above. I'd be happy to do that.

Yes, `exp` and `log` were not invariant because we chose to separate Riemannian and group operations into two categories and on group manifolds the Riemannian operations were operating on the manifold, with possibly non-invariant metric. `exp` and `log` can indeed be taken from the group structure but back when the decision to pass through was made, we didn't have Cartan-Schouten connections figured out yet. I'm open to changing this but as I said, it would be a bigger project. Right now to have an invariant `exp` and `log` you are supposed to use the `ConnectionManifold` type, which also selects the desired parallel transport (which is in general not unique for a group).

I'd say the current state is consistent enough to keep for a while until we have time to rethink the Riemannian/connection manifold operations we want on groups.

> 2\. As far as speed is concerned, `exp` is always an expensive operation, and I doubt that the overhead of function calls makes any difference. Am I wrong to assume that?

Actually it's not always expensive. On SE(3) with the old tangent vector representation we got down to about 50 ns using StaticArrays. Trait dispatch cost can easily reach an order of magnitude more. It's low enough that we can have it as a default but not low enough to never care about it.

> 3\. The only optimized `exp` that I can find in the code is that of `GeneralLinear`. It comes from a choice of left invariant metric, and I'm pretty sure it is not the same as the invariant exponential in `groups/connections.jl`. (So that optimized exponential is _not_ invariant, unless I am mistaken) Have I missed any other optimized `exp` implementation?

GL(N) in general does not admit a bi-invariant metric so the choice described in its docstring sounds fair to me. If the implementation doesn't match the docstring please make an issue.

Additionally we have some optimized implementations for `GeneralUnitaryMatrices`, SE(2) and SE(3).

Reply:
The invariant exponential for all Lie groups (regardless of tangent vector representation) is that one:
```julia
exp(G, p, X) = compose(G, p, exp_lie(G, inverse_translate_diff(G, p, p, X, LeftForwardAction())))
```

This implementation is **canonical** (i.e., the only invariant one) and works on any Lie group. I think that it should be the "reasonable default". 
(It does *not* depend on the choice of Cartan–Schouten parallel transport, which cannot be canonical). 

---

On SE(N) the previous implementations of `exp` were *not invariant*. This is not acceptable in robotics (the orientation of the robot at identity is arbitrary, like a choice of origin). (For SE(3), I do not know if it was invariant or not)

On GL(N), the canonical implementation above is invariant and canonical even though there are no bi-invariant metrics (and no invariant connections either).

As to `GeneralUnitaryMatrices`, I believe it is in fact an optimized implementation of `exp_lie`, but I may be wrong? Anyhow, it is a *manifold* exponential, not a *group* exponential. It should be the difference between `GeneralUnitaryMatrics` (the manifold) and `GeneralUnitaryMultiplicationGroup` (the group).

---

So I guess we fundamentally disagree for now: I would like to see the invariant exponential implementation by default for all Lie groups (even if it can be specialized), and, for now, you do not. Is that a correct assessment?

Reply:
> On SE(N) the previous implementations of exp were not invariant. This is not acceptable in robotics (the orientation of the robot at identity is arbitrary, like a choice of origin). (For SE(3), I do not know if it was invariant or not)

> So I guess we fundamentally disagree for now: I would like to see the invariant exponential implementation by default for all Lie groups (even if it can be specialized), and, for now, you do not. Is that a correct assessment?

As I said, in the current Manifolds.jl you are supposed to use `ConnectionManifold(SpecialEuclidean(3), CartanSchoutenMinus())` (or some other connection) if you want `exp` to be the invariant one. You don't have to use `exp` on `SpecialEuclidean`. I'm open to changing the default of `exp` and `log` on groups but not by just patching these two functions but by not passing any metric-related functions to the underlying manifold. So from traits of groups with non-bi-invariant metrics you'd have to remove `IsEmbeddedManifold`, `IsDefaultMetric`, `active_traits(f, M.manifold, args...)` and most likely `IsExplicitDecorator` and see if everything that's supposed to work still works. Then check performance regressions and finally update documentation. It would probably take me a few days to sort it out and there is an easy workaround so I can't justify to myself that it's a productive way to spend time.

In the meantime we can maybe have an alias for `ConnectionManifold(SpecialEuclidean(N), CartanSchoutenMinus())`?

> As to GeneralUnitaryMatrices, I believe it is in fact an optimized implementation of exp_lie, but I may be wrong? Anyhow, it is a manifold exponential, not a group exponential. It should be the difference between GeneralUnitaryMatrics (the manifold) and GeneralUnitaryMultiplicationGroup (the group).

I thought these have a bi-invariant metric?

Reply:
Just to be perfectly clear: in general, on any manifold, `exp` only depends on a **connection modulo its torsion**. (Example: all the Cartan–Schouten connections are the same except for their torsion; any of them defines the same exponential). To be clearer: a metric is not needed at all for geodesics (and thus exp) to be defined.

So, what I am suggesting is that all groups are by default endowed with one of the Cartan–Schouten connections.

If I had to choose one, it would be `CartanSchoutenZero`, since it has zero torsion.

So, any of the following would be acceptable solutions for me (it's ok if you don't want any of them of course!):
1. *No exponential at all on groups.* You'd have to explicitly choose a connection. This way, mistakes are not possible.
2. *All groups are endowed with one of the Cartan–Schouten connection by default* (preferably `CartanSchoutenZero`, although it doesn't matter for `exp`). Is it easy to do?
3. *All groups have `exp` and `log` defined invariantly*. That was my choice, it's easy to do, and all the tests were passing.

For performance reason, you do not want 3., but it's the easiest to implement, I guess?

(The existence of a bi-invariant metric is irrelevant, in my opinion. Some groups have it some don't, but they all have an invariant connection, any of which gives the same exponential.)

> In the meantime we can maybe have an alias for ConnectionManifold(SpecialEuclidean(N), CartanSchoutenMinus())?

Yes, but can this be done for all groups, as suggested above in 2.? Something like that (alias or default) would be an acceptable solution, I think.

Reply:
I'm OK with either (2) or (3) but neither of them is easy to do right I fear. The problem with (3) is performance and that nothing related the the Riemannian structure of the underlying manifold should work. We don't have test for either of those things currently and just because we don't doesn't mean we don't care about them. If tests are passing with your change, it means there is not enough tests.

Keep in mind that the current design of Manifolds.jl was centered around Riemannian manifolds and groups were seen as a useful addition. Most of the things regarding Lie group theory you mention here are things I wasn't aware of at the time the design was made but I've learned about them around the time I've added Cartan-Schouten connections, which I did in a way consistent with existing design choices. I can see the drawbacks but at least it has some consistent logic and just replacing `exp` and `log` would violate that logic.

> > In the meantime we can maybe have an alias for ConnectionManifold(SpecialEuclidean(N), CartanSchoutenMinus())?
> 
> Yes, but can this be done for all groups, as suggested above in 2.? Something like that (alias or default) would be an acceptable solution, I think.

Sure, we can have for example `SpecialEuclideanZero` etc. aliases for all groups.

Reply:
>  If tests are passing with your change, it means there is not enough tests.

Any concrete idea of a test missing?

I mean, `exp` was defined for some groups before, and everything was working? My change does not change very much what was already in place (see below).

> just replacing exp and log would violate that logic.

But `exp` and `log` were *already replaced* in many cases. For instance, the `exp` for `SpecialEuclidean` or `GeneralLinear` are not specifically related to a metric, and nothing wrong happens.

---
I don't really understand the concern here, but I think that at least, to make it very clear for the user, there should be a table explaining from which zero-torsion connection each `exp` is based on. Here are the ones that I could figure out so far:

| Group | Zero torsion connection | Invariant |
| --- | --- | --- |
| `SemidirectProductGroup{<:Any, <:LeftInvariantRepresentation}`  | `CartanSchoutenZero` (explicitly in `semidirect_product_group.jl:296` | ✅ |
| `SemidirectProductGroup{<:Any, <:HybridTangentRepresentation}` | Metric connection from the product of implicit metrics on `N` and `H` | ❌ |
| `GeneralLinearGroup` | Metric connection from the left invariant metric induced from the standard basis on the Lie algebra | ❌ |
| `GeneralUnitaryMultiplicationGroup` | `CartanSchoutenZero` (explicitly as such in `GeneralUnitaryMatrices.jl:225`)  | ✅ |
| `HeisenbergGroup` | `CartanSchoutenZero`? (probably defined explicitly as such in `heisenberg.jl:220`) | ✅? |
| `SpecialLinearGroup` | Same as `GeneralLinear` | ❌ |

Let me know if I have forgotten a group.

As you see, in these cases, `exp` is redefined, without reference to a connection or a metric, and everything is just fine.

Reply:
> Any concrete idea of a test missing?
> 
> I mean, `exp` was defined for some groups before, and everything was working? My change does not change very much what was already in place (see below).
> 
> >

Things like "call X executes below a certain time threshold" or "all metric/connection functions on manifold X are either consistent or throw an error".

> I mean, `exp` was defined for some groups before, and everything was working? My change does not change very much what was already in place (see below).

`exp` was meant to be the Riemannian or connection exponential, universally in all Manifolds.jl, not the group one. The group one is `exp_lie`.

> But `exp` and `log` were _already replaced_ in many cases. For instance, the `exp` for `SpecialEuclidean` or `GeneralLinear` are not specifically related to a metric, and nothing wrong happens.

As far as I am aware `SpecialEuclidean` was replaced with a faster implementation of the Riemannian logic and `exp` of `GeneralLinear` should implement the Riemannian exponential. If either of these are not true, it's a bug, not a feature.

> I don't really understand the concern here, but I think that at least, to make it very clear for the user, there should be a table explaining from which zero-torsion connection each `exp` is based on. Here are the ones that I could figure out so far:

I see, we just didn't care until this point whether `exp` is group-invariant or not. So we have no such table and the interface wasn't designed with consistency in this part in mind. And I thought semidirect products are the only case where we don't have left-invariance?

I'm not sure why you marked the GL and SL `exp` as not invariant? I thought it should be left-invariant.

Remember that existence of a specialized implementation doesn't always mean that it's supposed to change logic, in some cases it's just for performance.

Reply:
> I'm not sure why you marked the GL and SL exp as not invariant? I thought it should be left-invariant.

What I mean by invariance is that it is completely canonical, uniquely defined from `exp_lie` by invariance. The math definition is
```math
\exp_p(pX) = p \exp(X)
```
where the second `exp` is the Lie exponential `exp_lie`.

The `GL` exponential as it is implemented comes from a *non-invariant* metric connection coming from a *left-invariant* metric (which is not canonical, it depends on the basis). It is confusing.

> existence of a specialized implementation doesn't always mean that it's supposed to change logic, in some cases it's just for performance.

My point exactly: these specialized implementations *actually change the logic*. Shouldn't that be clearly documented for the users?

Reply:
> The `GL` exponential as it is implemented comes from a _non-invariant_ metric connection coming from a _left-invariant_ metric (which is not canonical, it depends on the basis). It is confusing.

Yes, currently the Riemannian metric or connection structure is fundamental for Manifolds.jl, not the group one. We can change that but it's not a small project. Maybe seeing the library from this point of view would make it easier to understand.

We wanted to have a Riemannian metric on GL(n) and IIRC we didn't find any that gives invariant geodesics. We preferred to have a Riemannian metric to having invariant exponential map with a non-metric connection.

> My point exactly: these specialized implementations _actually change the logic_. Shouldn't that be clearly documented for the users?

Could you be more specific which methods change the logic? Anyway, if some implementation changes the logic, we should change those methods so that the no longer change it. This is the fundamental part. They should not change the logic. `exp` and `log` should always come from some connection, their invariance to group action must follow from the choice of such connection.

Again, in Manifolds.jl `exp` is **not** the Lie exponential and where it is, it is a coincidence. Maybe we should make it more clear in documentation? I could surely write about it if you can see a good place for it, or maybe there should be even a new tutorial about groups.

Reply:
The main point we did introduce `exp_lie` was _because_ `exp` was always meant to be one for a Riemannian metric on the manifold.

> we should change those methods so that the no longer change it. 

I agree with that.

The more I read these discussions the more I think it would be clearer to really make the distinction more clear between Riemannian manifolds (from which we sometimes only keep the set) and Lie groups build upon that.

>  there should be even a new tutorial about groups.

I‘d love to have that, but miss the capacities currently myself.

Reply:
> The main point we did introduce `exp_lie` was _because_ `exp` was always meant to be one for a Riemannian metric on the manifold.

Yes, `exp_lie` and `log_lie` are the functions that guarantee invariance. Not `exp` and `log`.

Reply:
Thank you for all your answers!

I think a summary of our positions is this:
*My position*: `exp` should be defined from the natural connection on Lie groups (because many users will assume this)
*Your position*: `exp` should be defined from *some arbitrary connection* (sometimes the natural one by coincidence, sometimes not; aka sometimes `exp` is invariant, sometimes not)

Is it an accurate statement?

---

Here are the two definitions that guarantee that `log` and `exp` are invariant (i.e., defined from the natural connection). They are **fast** with left invariant representation, and work on all Lie groups:
```julia
log(G, p, q) = translate_diff(G, p, Identity(G), log_lie(G, compose(G, inv(G, p), q)), LeftForwardAction())
exp(G, p, X) = compose(G, p, exp_lie(G, inverse_translate_diff(G, p, p, X, LeftForwardAction())))
```
---
> We can change that but it's not a small project. 

I don't think so: I did it and it worked since all the tests were passing? Would you have a *very specific* test function that wouldn't pass?

> Could you be more specific which methods change the logic? Anyway, if some implementation changes the logic, we should change those methods so that the no longer change it. 

"Change the logic" means for me: not being invariant. **Specifically**, on `GL`, on `SL` and on semi-direct product with hybrid representation (see table above).

> This is the fundamental part. They should not change the logic. `exp` and `log` should always come from some connection, 

They all do! All the `exp` implemented come from *some* connection (see table above). But on groups, I think that they should depend on *the natural* connection, instead of some arbitrary other one.

If "Change the logic" means having `exp` not being defined from a connection, this is *not* what I'm suggesting.

> their invariance to group action must follow from the choice of such connection.

**There is no choice of a connection**: if `exp` is to be invariant, it **must** come from the natural connection.

> Again, in Manifolds.jl `exp` is **not** the Lie exponential and where it is, it is a coincidence. 

I'm not sure I'm following: these are two different concepts since `exp_lie` maps the Lie algebra to the group, `exp` maps any tangent space to the group. But the implementation above defines `exp` from `exp_lie` in a canonical way on any Lie group.

Reply:
Currently the central object is a manifold. That will stay that way, see how the package is called.
We interpret a Lie group here as “wrapping” the Manifold in some thing that adds a group action.
`exp` is a function (in  the package) that belongs to the manifold. So we respect that and pass that down.
But “adding a group action” currently does not change the metric of the Manifold.
Or in other words: `exp` on a manifold changes only if you explicitly change the metric (see `MetricManifold` but not implicitly just because you add a group action.

That is a purely design decision.

You arguments are completely valid and the default exp on a Lie group sounds nice, but that would be something I would consider a good design decision in a package called `LieGroups.jl`. I am not even against doing such a package, I think that would be nice, taking all designs we have here, taking `Manifolds.jl` as a dependency and doing such a package. I even proposed such an Idea in an earlier longer discussion.

That does not mean your arguments are invalid, they are concise and nice, they would just change one of the main design decisions of this package with probably huge implications on the total code base.

I just do not have the capacity for starting another package. The ones I have (questions, inquiries, improvements, bug fixes) keep me busy enough.
Maybe that changes some time, but probably not soon.

Reply:
Overall I am not yet so sure what you want to achieve here through this longish discussion:

1) Do you want to convince us that we are stupid? I feel your tone if often quite agressive and that at least for me sounds like you would like to convince us, that we are stupid – and maybe should stop developing this package?

2) Since we alreaady agreed to all your metric agruments – do you want us to rewrite the whole package? Before we do that I would really recommend to start Liegroups.jl and “do Lie Groups right” following your arguments and use Manifolds.jl as a kind of backend (for example for point checks or such).
I am even on board helping with that a bit, setting everything up here in JuliaManifolds, CI / docs / ... / review PRs.

Reply:
Ah just when I wanted to call it a day, I had an idea: We could add a keyword argument to `LIeGroup` that says: I want to have the Olivier-approach (name to be found still)
Then the manifold that is passed would be wrapped in a MetricManifold with the corresponding metric (the one that is the best one in Oliviers sense) and on that dispatch the default from Olivier would work.

If we later notice that the approach of changing the metric is for some reason better, it would just mean to change the default in the constructor (from not wrapping to wrapping in the nicest-and-best-of-all-for-Lie metric). That would by then be breaking but until then we have both solution relatively easily in code.

Reply:
> I think a summary of our positions is this: _My position_: `exp` should be defined from the natural connection on Lie groups (because many users will assume this) _Your position_: `exp` should be defined from _some arbitrary connection_ (sometimes the natural one by coincidence, sometimes not; aka sometimes `exp` is invariant, sometimes not)
> 
> Is it an accurate statement?

It isn't entirely accurate; the connection is currently arbitrary but I'm open to changing that, just in a consistent manner.

> Here are the two definitions that guarantee that `log` and `exp` are invariant (i.e., defined from the natural connection). They are **fast** with left invariant representation, and work on all Lie groups:

I got that point but it appears you don't understand my concerns at all.

> I don't think so: I did it and it worked since all the tests were passing? Would you have a _very specific_ test function that wouldn't pass?

Sure. Start with making `inner`, `norm`, `get_vector` and `get_coordinates` (for orthonormal and orthogonal bases) throw `MethodError` on every Lie group where `exp` and `log` does not come from a Levi-Civita connection of some metric. That's not all but it would be a good start. Also make sure the current non-invariant `exp` and `log` on GL and SL is usable in some (consistent) way.

> "Change the logic" means for me: not being invariant. **Specifically**, on `GL`, on `SL` and on semi-direct product with hybrid representation (see table above).

This is not at all what I refer to "change the logic". You made an incorrect assumption about how Manifolds.jl operates and refuse to adjust your expectations despite my detailed explanations.

> They all do! All the `exp` implemented come from _some_ connection (see table above). But on groups, I think that they should depend on _the natural_ connection, instead of some arbitrary other one.
> 
> If "Change the logic" means having `exp` not being defined from a connection, this is _not_ what I'm suggesting.

Whenever I was talking about changing the logic, I meant `exp` not corresponding to the connection chosen for a group. This is the fundamental logic of `exp`. If you don't like it, don't use `exp`. This is one part where I see no point in changing anything about Manifolds.jl: `exp` must come from a consistent choice of connection. We can change which connection it is but invariance is not a requirement for `exp`.

> **There is no choice of a connection**: if `exp` is to be invariant, it **must** come from the natural connection.

There is some literature on non-invariant connections. Some people find them useful. I don't see why we can't support such use cases, specially given that some of it is already implemented.

> I'm not sure I'm following: these are two different concepts since `exp_lie` maps the Lie algebra to the group, `exp` maps any tangent space to the group. But the implementation above defines `exp` from `exp_lie` in a canonical way on any Lie group.

Maybe a good way to resolve this disagreement would be to introduce `exp_inv` and `log_inv` (invariant exp and invariant log) which would be implemented in the exact way you described?

Reply:
I think `exp_inv` and `log_inv` would also be good ideas – they would be a bit more specific than wrapping the manifold in a metric (as I proposed in my last comment) and dispatch with `exp` on that. I would be fine with both. Maybe the wrapping idea avoids introducing (yet) another function name.

Reply:
I get it now (I think): you want `exp` to be defined from a **metric** connection.

If `exp` is defined from a nonmetric connection, you are quite sure that something is going to break (because Lie groups are built on top of metric manifolds).

> We could add a keyword argument to `LIeGroup` that says: I want to have the Olivier-approach (name to be found still) Then the manifold that is passed would be wrapped in a MetricManifold with the corresponding metric (the one that is the best one in Oliviers sense) and on that dispatch the default from Olivier would work.

Yes, that would be nice, but it may not be possible, since the invariant connection is not necessarily metric (example: GL, semidirect products).

> It isn't entirely accurate; the connection is currently arbitrary but I'm open to changing that, just in a consistent manner.

For a metric connection, there is no consistent solutions, your solution is as good as any other. If you agree with (possibly) nonmetric connections, the invariant one is the only consistent choice across all Lie groups.

> You made an incorrect assumption about how Manifolds.jl operates and refuse to adjust your expectations despite my detailed explanations.

I guess I did. Sorry for that.
 
> Whenever I was talking about changing the logic, I meant `exp` not corresponding to the connection chosen for a group. This is the fundamental logic of `exp`. If you don't like it, don't use `exp`. This is one part where I see no point in changing anything about Manifolds.jl: `exp` must come from a consistent choice of connection. We can change which connection it is but invariance is not a requirement for `exp`.

You mean a metric connection. Tu sum it up, you prefer a *metric noninvariant* connection over an *invariant nonmetric* connection.

This is our disagreement in a nutshell.

> Maybe a good way to resolve this disagreement would be to introduce `exp_inv` and `log_inv` (invariant exp and invariant log) which would be implemented in the exact way you described?

Yes, that would be very nice. I think those two functions deserve to be defined somewhere.

---

>  1. Do you want to convince us that we are stupid? I feel your tone if often quite agressive 

It's just a misunderstanding, a disagreement. I like to understand things. I didn't and it's my fault, I'm slow, it makes me sound aggressive. But I think I get it now. Nobody is stupid here, or maybe we're all stupid in the face of maths.

>   I would really recommend to start Liegroups.jl and “do Lie Groups right” following your arguments and use Manifolds.jl as a kind of backend (for example for point checks or such).
>        I am even on board helping with that a bit, setting everything up here in JuliaManifolds, CI / docs / ... / review PRs.

I really appreciate that. I kind of already started actually. I'll let you know how it goes.

Tell me if we finally understand each other (i.e.: you only want `exp` defined from a metric connections, otherwise you know something will break). If we do, we can resolve this conversation.

Reply:
I am fine with us all being stupid ;)
Sorry your tone (and repetition as it came across to me) was getting to me as agressive – thanks for clarifying that this was not meant like that.

My main problem until now was, that to me this thread slowly read like we are repeating arguments again and again.
But if you are fine with the `exp_inv`/`lie_inv` approach, that is fine with me as well. And yes, I think we by now understand each other, what we want from exp and why that is different fro yours.

For the `LieGroups.jl` idea; I think a nice thing would also be that we do not “overcrowd” `Manifolds.jl` but modularise features a bit. If you want and have done some work on that, as I said, I am very happy to help with all technical details in setting that up; I would prefer that in JuliaManifolds/ – just to not have a package in a persons namespace – but you would be admin of that package as well of course.

Reply:
> I get it now (I think): you want `exp` to be defined from a **metric** connection.

Yes, we chose the connections for GL and SL because we prefer the defaults to be metric non-invariant connections over non-metric invariant connections.  Not all `exp` methods come from a metric connection but we generally prefer metric connections because they give more rich structure required by certain optimization or statistical algorithms; moreover, not all manifolds are Lie groups.

> You mean a metric connection. Tu sum it up, you prefer a _metric noninvariant_ connection over an _invariant nonmetric_ connection.
> 
> This is our disagreement in a nutshell.
> 

Yes, that's correct.

> Yes, that would be very nice. I think those two functions deserve to be defined somewhere.

OK, great, then I will add them in my PR and mention in the new tutorial.

> > 1. Do you want to convince us that we are stupid? I feel your tone if often quite agressive
> 
> It's just a misunderstanding, a disagreement. I like to understand things. I didn't and it's my fault, I'm slow, it makes me sound aggressive. But I think I get it now. Nobody is stupid here, or maybe we're all stupid in the face of maths.

For me developing this library is a great learning opportunity. I'm not a mathematician, and primarily see various branches of differential geometry and Lie groups as a nice language for generalizing classical numerical algorithms. I'm sometimes wrong, but when someone shows me why I'm wrong I can learn something new. There are definitely dozens of people who would do much better job developing JuliaManifolds but they are too busy with more important stuff.

I'm happy when people find JuliaManifolds useful and I generally consider any suggestions regarding interface and functionality of libraries. However, it's impossible to satisfy everyone and we have to set some priorities.

I generally don't mind the tone of messages I get about JuliaManifolds, either yours and from other people. I know it's sometimes a real hassle to express in a positive way, even more so if someone is not a native speaker.

> Tell me if we finally understand each other (i.e.: you only want `exp` defined from a metric connections, otherwise you know something will break). If we do, we can resolve this conversation.

Yes, we prefer metric connections if they are available; I think we've managed to resolve the misunderstanding here finally :slightly_smiling_face: .

> For the `LieGroups.jl` idea; I think a nice thing would also be that we do not “overcrowd” `Manifolds.jl` but modularise features a bit. If you want and have done some work on that, as I said, I am very happy to help with all technical details in setting that up; I would prefer that in JuliaManifolds/ – just to not have a package in a persons namespace – but you would be admin of that package as well of course.

Not-so-fun fact: https://github.com/yuehhua/LieGroups.jl already exists, is registered and looks abandoned. I think it would be fine to develop Lie groups in a separate package but currently we have a bunch of functionality here and maybe some part of it should be moved there?

Reply:
Oh, I did not see that package, well, if ours gets a bit more serious we could reach out to hopefully take over the name, if that is abandoned. It also seems to not cover many (only one?) Lie group.